src/python/xml_parser-py.c

```diff
@@ -792,3 +790,271 @@ py_xml_parse_main_metadata_together(G_GNUC_UNUSED PyObject *self, PyObject *args
+        self->pkg_iterator = NULL;
+        self->cbdata = malloc(sizeof(CbData));
+    }
+    return (PyObject *)self;
+}
+
+PyDoc_STRVAR(pkg_iterator_init__doc__,
+             "PkgIterator object\n\n"
+             ".. method:: __init__()\n\n"
+             "    Default constructor\n");
+
+static int
+pkg_iterator_init(_PkgIteratorObject *self, PyObject *args, PyObject *kwargs)
+{
+    char *primary_path;
+    char *filelists_path;
+    char *other_path;
+    PyObject *py_newpkgcb, *py_warningcb;
+    GError *tmp_err = NULL;
+    static char *kwlist[] = {"primary", "filelists", "other", "newpkgcb",
+                             "warningcb", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwargs, "sssOO:pkg_iterator_init", kwlist,
+                                     &primary_path, &filelists_path, &other_path, &py_newpkgcb,
+                                     &py_warningcb)) {
+        return -1;
+    }
+
+    if (!primary_path || !filelists_path || !other_path) {
+        PyErr_SetString(PyExc_TypeError, "file paths must be provided");
+        return -1;
+    }
+
+    if (!PyCallable_Check(py_newpkgcb) && py_newpkgcb != Py_None) {
+        PyErr_SetString(PyExc_TypeError, "newpkgcb must be callable or None");
+        return -1;
+    }
+
+    if (!PyCallable_Check(py_warningcb) && py_warningcb != Py_None) {
+        PyErr_SetString(PyExc_TypeError, "warningcb must be callable or None");
+        return -1;
+    }
+
+    if (self->pkg_iterator) // reinitialization by __init__()
+        cr_PkgIterator_free(self->pkg_iterator, &tmp_err);
+    // TODO: correct thing to do with error?
+    if (tmp_err) {
+        nice_exception(&tmp_err, NULL);
+        return -1;
+    }
+
+    Py_XINCREF(py_newpkgcb);
+    Py_XINCREF(py_warningcb);
+
+    cr_XmlParserNewPkgCb ptr_c_newpkgcb = NULL;
+    cr_XmlParserWarningCb ptr_c_warningcb = NULL;
+
+    if (py_newpkgcb != Py_None)
+        ptr_c_newpkgcb = c_newpkgcb;
+    if (py_warningcb != Py_None)
+        ptr_c_warningcb = c_warningcb;
+
+    self->cbdata->py_newpkgcb = py_newpkgcb;
+    self->cbdata->py_pkgcb = NULL;  // TODO: What should go here?
+    self->cbdata->py_warningcb = py_warningcb;
+    self->cbdata->py_pkgs = PyDict_New();
+
+    self->pkg_iterator = cr_PkgIterator_new(
+        primary_path, filelists_path, other_path, ptr_c_newpkgcb, self->cbdata, ptr_c_warningcb, self->cbdata, &tmp_err);
+
+    // TODO: correct?
+    if (tmp_err) {
+        nice_exception(&tmp_err, NULL);
+        return -1;
+    }
+
+    if (self->pkg_iterator == NULL) {
+        PyErr_SetString(CrErr_Exception, "PkgIterator initialization failed");
+        return -1;
+    }
+
+    return 0;
+}
+
+static void
+pkg_iterator_dealloc(_PkgIteratorObject *self)
+{
+    GError *tmp_err;
+    if (self->pkg_iterator) {
+        cr_PkgIterator_free(self->pkg_iterator, &tmp_err);
+    }
+    if (self->cbdata) {
+        // Py_XDECREF(self->cbdata->py_newpkgcb);
+        // Py_XDECREF(self->cbdata->py_warningcb);
```

Comment:
I've commented out these two lines, uncommenting either of them leads to segfault on the added test

Reply:
I see, `Py_XDECREF` can take `NULL`s so the problem is we don't always initialize `self->cbdata` it is just `malloced` so there are random data (random addresses) and that doesn't go well.

I think the best fix would be to change the line 
```c
self->cbdata = malloc(sizeof(CbData));
```
to
```c
self->cbdata = g_malloc0(sizeof(CbData));
```